### ISSUE TITLE
Symfony DebugClassLoader causes deprecation notices

### ISSUE THREAD

**Issue by jvdlaar** [amphp/amp#257]
When using Symfony 4.2 I'm getting these deprecation notices:
```
14:26:52 INFO      [php] User Deprecated: The "Amp\Loop\NativeDriver::onSignal()" method will require a new "callable (string $watcherId" argument in the next major version of its parent class "Amp\Loop\Driver", not defining it is deprecated. ["exception" => ErrorException { …}]
14:26:52 INFO      [php] User Deprecated: The "Amp\Coroutine::onResolve()" method will require a new "callable(\Throwable|null $reason" argument in the next major version of its parent class "Amp\Promise", not defining it is deprecated. ["exception" => ErrorException { …}]
14:26:52 INFO      [php] User Deprecated: The "Amp\Internal\PrivatePromise::onResolve()" method will require a new "callable(\Throwable|null $reason" argument in the next major version of its parent class "Amp\Promise", not defining it is deprecated. ["exception" => ErrorException { …}]
14:26:52 INFO      [php] User Deprecated: The "Amp\Success::onResolve()" method will require a new "callable(\Throwable|null $reason" argument in the next major version of its parent class "Amp\Promise", not defining it is deprecated. ["exception" => ErrorException { …}]
14:26:52 INFO      [php] User Deprecated: The "Amp\Failure::onResolve()" method will require a new "callable(\Throwable|null $reason" argument in the next major version of its parent class "Amp\Promise", not defining it is deprecated. ["exception" => ErrorException { …}]
```

I think this is caused because symfony doesn't understand this:
```
* @param callable(\Throwable|null $reason, $value) $onResolved `$reason` shall be `null` on
*     success, `$value` shall be `null` on failure.
```

I'm not really sure if this is because above annotation is wrong or symfony is wrong tho.

**Comment by kelunik** [comment]
@jvdlaar This is caused by Symfony not supporting typed callables (which is reasonable), but I'm not sure why that causes warnings like that. Please open an issue there and reference this issue.

**Comment by stof** [comment]
The DebugClassLoader has a feature to allow adding new optional methods in the signature of an interface without adding them for real (which would be a BC break for classes implementing the interface): by adding the param in the phpdoc of the interface, the DebugClassLoader will report a deprecation for child classes not adding this optional param when implementing the interface (which is perfectly valid in phpdoc, as it is optional). The goal being that the interface can add this param in the next major version, and the code is already compatible with it (if the deprecation warning was resolved).

Here, it indeed fails because this `@param callable(\Throwable|null $reason, $value) $onResolved` syntax is not supported (it is currently detected as a `$reason` parameter, with the `callable(\Throwable|null` type).
Is this notation even valid phpdoc, or would it also confuse IDEs and static analyzers ?

**Comment by bwoebi** [comment]
No it isn't. There however is no alternate syntax in phpdoc for describing callable signatures, so there's that...

**Comment by stof** [comment]
well, an option could be to make that valid phpdoc, by moving the doc for the signature of the callable to the description of the argument:

```
@param callable $onResolved  The callable should
      have (\Throwable|null $reason, $value) as a signature. `$reason` shall
      be `null` on success, `$value` shall be `null` on failure.
```

**Comment by jvdlaar** [comment]
Seems like symfony released a fix for this: https://github.com/symfony/symfony/commit/3659e32279e28b918f406deb50f94efd0f819467 so I think this will be fixed with that. I'll try to verify this.

**Comment by stof** [comment]
Note that this "fix" will only be part of Symfony 4.3 (due in May), as it was considered a feature, not a bugfix (if that phpdoc was valid, we would have considered it a bugfix)

**Comment by kelunik** [comment]
As phpdoc doesn't have a formal definition as a finished spec anywhere, I'd consider it a bugfix.
